clippy_lints/src/unnecessary_self_imports.rs

```diff
@@ -0,0 +1,177 @@
+                        snippet_with_applicability(cx, ident.span, "..", &mut applicability)
+                    )
+                };
+                span_lint_and_sugg(
+                    cx,
+                    UNNECESSARY_SELF_IMPORTS,
+                    path.span,
+                    "import ending with `self`",
+                    "consider omitting `::{self}`",
+                    snippet,
+                    applicability,
+                );
+            }
+        }
+    }
+}
+
+fn is_self_import(cx: &LateContext<'_>, current_import: &Item<'_>, current_import_path: &Path<'_>) -> Option<Ident> {
+    let definitions = cx.tcx.hir().definitions();
+    let mut amt = 0;
+    let mut ident: Option<Ident> = None;
+
+    for def in definitions.iter_local_def_id() {
+        if_chain! {
+            if let Some(Node::Item(item)) = cx.tcx.hir().get_if_local(def.to_def_id());
+            if let ItemKind::Use(path, kind) = item.kind;
+            if matches!(kind, UseKind::Single | UseKind::Glob);
+            if current_import.span.contains(item.span);
+            then {
+                amt += 1;
+                if amt > 1 { return None; }
+
+                if_chain! {
+                    if current_import_path.segments.len() == path.segments.len();
+                    let current_import_last = &current_import_path.segments[current_import_path.segments.len() - 1];
+                    let item_last = &path.segments[path.segments.len() - 1];
+                    if current_import_last.ident == item_last.ident;
+                    then {
+                        ident = Some(item.ident);
+                    }
+                }
+            }
+        }
+    }
+    ident
+}
+
+fn mod_contains_item(cx: &LateContext<'_>, def_id: DefId, ident: Ident) -> bool {
+    if let Some(Node::Item(node)) = cx.tcx.hir().get_if_local(def_id) {
+        if let ItemKind::Mod(Mod { item_ids, .. }) = &node.kind {
+            for item in item_ids.iter() {
+                if_chain! {
+                    if let Some(Node::Item(node)) = cx.tcx.hir().get_if_local(item.def_id.to_def_id());
+                    if let VisibilityKind::Public = node.vis.node;
+                    if node.ident == ident;
+                    if matches!(node.kind, ItemKind::Fn(..) | ItemKind::Const(..) | ItemKind::Static(..));
+                    then { return true; }
+
+                }
+            }
+        }
+    } else {
+        for item in cx.tcx.item_children(def_id).iter() {
+            if_chain! {
+                if item.ident == ident;
+                if matches!(item.res, Res::Def(DefKind::Fn | DefKind::Const | DefKind::Static, _));
+                then { return true; }
+            }
+        }
+    }
+    false
+}
+
+fn item_is_in_scope(cx: &LateContext<'_>, ident: Ident, current_import: Span) -> bool {
+    let definitions = cx.tcx.hir().definitions();
+    let mut outer_mod: Option<Span> = None;
+
+    for def in definitions.iter_local_def_id() {
```

Comment:
Could you add some comments what `mod_contains_item` and `item_in_scope` do? Iterating through **all** definitions in a crate seems a little overkill to me for this lint.